llvm/lib/CodeGen/MachineBasicBlock.cpp

```diff
@@ -1403,7 +1403,15 @@ bool MachineBasicBlock::canSplitCriticalEdge(
   const MachineFunction *MF = getParent();
   // Performance might be harmed on HW that implements branching using exec mask
   // where both sides of the branches are always executed.
-  if (MF->getTarget().requiresStructuredCFG())
+  // However, if `Succ` is a loop header, splitting the critical edge will not
+  // break structured CFG.
+  auto SuccIsLoopHeader = [&]() {
+    if (MLI)
+      if (MachineLoop *L = MLI->getLoopFor(Succ); L && L->getHeader() == Succ)
+        return true;
+    return false;
```

Comment:
```suggestion
    if (MLI)
      const MachineLoop *L = MLI->getLoopFor(Succ); 
      return L && L->getHeader() == Succ;
```

This also doesn't need to be a lambda